Debug symbols for release builds

posborne opened the pull request:
Feedback appreciated on this one.  My first shot at this was to just include debug symbols in the output ELF.  That in and of itself, however, still did not result in me being able to do anything useful as far as debugging the blink example (which I happened to be using for my testing).  I think this is because of the optimizations (lots of inlining for sure).

Debugging works well with the "dev" profile which is defaulted to by cargo when `--release` is dropped.

posborne commented on Makefile.in:
```diff
@@ -1,17 +1,26 @@
-OBJDUMP=@OBJDUMP@
+STRIP ?= @STRIP@
+OBJCOPY ?= @OBJCOPY@
+OBJDUMP ?= @OBJDUMP@
```
Changing these to `?=` could be nice but could also cause problems if one has a polluted environment.  Thoughts?

mcoffin replied:
I like it. I get more frustrated when i set `CC` and it doesn't effect a Makefile than when I forget I have it set and it compiles for the wrong target until I unset it.

mcoffin commented on Makefile.in:
```diff
@@ -1,17 +1,26 @@
+	PROFILE_OPTION = --release
+	PROFILE = release
+endif
+
```
:+1:

posborne committed:
build: allow user to build with dev/debug profile

The user can now invoke builds like the following:

    $ make DEBUG=true EXAMPLE_NAME=blink

This will build artifacts with the dev profile that includes
debug symbols and eliminates optimization that would normally be
included.  This is necessary to get a reasonable debugging experience
in many cases.

A few options were also modified in order to allow for the user
to override them if desired.

mcoffin commented:
LGTM, but we're not using the lieutenant workflow, so pull requests to the master branch of forks don't make sense because the master branch of a fork means nothing except possibly to the owner of the fork. (PRs to current topic branches in forks, however, would make perfect sense (ex. #3)). You'll want to re-open this [upstream](/hackndev/zinc).

posborne commented:
Haha, oops.  I did not intend to open the PR against the fork.  Github must have noticed that I had done this previously and assumed this is what I wanted now.  Not sure.  Will open against upstream.
